distill/core/core_types.py

```diff
@@ -0,0 +1,3 @@
+from typing import Any
+
+JsonObject = dict[str, Any]
```

Comment:
We probably need a more robust type hint here.
```py
from typing import Union, List, Dict

# A JSON-serializable type can be:
# - a string, number, boolean, or None
# - a list of JSON-serializable types
# - a dictionary with string keys and JSON-serializable values

JSONSerializable = Union[
    str,
    int,
    float,
    bool,
    None,
    List['JSONSerializable'],  # Use the type alias recursively
    Dict[str, 'JSONSerializable']
]

# Example function using this type hint
def process_json(data: JSONSerializable) -> None:
    print(data)

# Example usage
example_data: JSONSerializable = {
    "name": "John",
    "age": 30,
    "is_member": True,
    "balance": 100.50,
    "friends": ["Alice", "Bob"],
    "metadata": {
        "height": 175.5,
        "active": False
    }
}

process_json(example_data)
```
You could also alias the json dict:
```py
JsonDict = Dict[str, 'JSONSerializable']
```
And re-use that.

Any is too permissive.

Reply:
Didn't even know you could do recursive definitions like that, thanks

Reply:
@EandrewJones if we allow JSONSerializable to be a other things besides Dict[str, something] then we lose the ability to pass the `data` object to our model with `schema(**data)`. 
![image](https://github.com/apache/flagon-distill/assets/73962982/ffb155cf-6bd4-4a28-9fa3-632c286c7fb3)
The `BaseModel` that our UserAle schema inherits from expects an input of keyword arguments corresponding to the underlying schema.

Reply:
How do you feel about something like this? 
![image](https://github.com/apache/flagon-distill/assets/73962982/37a75567-d6eb-4ba6-b2ca-e8f8b6b0f216)

Reply:
But that's not a proper definition for a JSONSerializable object.

What I specified should work, but then you change the type hint for `data` in `__init__`:
```py
def __init__(self, data: Union[str, JsonDict], schema=UserAleSchema):
```
The log should me a JSON serializable dict, but the type interface for _all_ JSONSerializable objects is more expansive.